Drafts/1NNN-SB.md

```diff
@@ -0,0 +1,267 @@
+
+## Description
+**Currently**: An template alias function parameter is not resolved to its aliased instance until after the function
+resolution stage. So, with the following code:
+```
+struct TemplateType(T) {}
+template aliasAlias(T) = TemplateType!T;
+void templateFunction(T)(TemplateAlias!T arg) {}
+
+void main()
+{
+    TemplateAlias!int inst;
+    templateFunction(inst); /* "cannot deduce function from argument types !()(TemplateType!int)" */
+}
+```
+We get the following error message:
+```
+template templateFunction cannot deduce function from argument types !()(TemplateType!int), candidates are:
+templateFunction(T)(TemplateAlias!T arg)
+```
+The compiler sees `TemplateAlias` and `TemplateType` as two different types, and so it can't match `TemplateType!int` against
+`TemplateAlias!T`. Although the compiler has resolved the type of `inst` to `TemplateType!int`, it has not resolved the 
+formal parameter type to be `TemplateType!T`.
+
+**New**:
+Full handling of the feature is complicated because a template (alias) has arbitrary pattern matching and rewrite abilities over types. For instance, an template alias:
+ * can accept type parameters that it never uses (e.g., `alias TemplateAlias(T) = int;`)
+ * can accept a type parameter and use it multiple times (e.g., `alias TemplateAlias(T) = TemplateType!(T,T);`)
+ * can infer a type parameter when given a complex type expression (e.g., `alias TemplateAlias(A: A*, B) = TemplateType!(A, B);`).
+ 
+Therefore, the crux of the proposal is to encode this pattern matching and type rewriting into a function, `Gen`, and invoke it at the right time during compilation. But it is not to define that pattern-matching or the type-rewriting.
+
+Specifically, if the type of the formal parameter of a template function is an instance of an template alias declaration, instantiate the type that the declaration
+aliases and that is the new type of the parameter. This is done by calling the function below in each parameter. This should be done before function calls are resolved.
+In pseudocode:
+```
+resolveAlias(ref parameterType) {
+    while (parameterType is templateInstance) {
+        for (arg in parameterType.paramMap.values) {
+            if (arg is templateInstance && arg.TD is templateAliasDeclaration) {
+                // Recursively resolve nested aliases
+                resolveAlias(arg);
+            }
+        }
+        // Assume that parameterType was generated using `Gen`
+        if (parameterType.TD is templateAliasDeclaration) {
+            aliasedType = TD.aliasedType;
+            if (aliasedType is templateInstance) {
+                newTD = aliasedType.TD;
+                newArgumentList = matchParams(aliasedType.argumentList, parameterType.paramMap);
+                parameterType = Gen(newTD, newArgumentList);
+                continue;
+            } else {
+                parameterType = aliasedType;
+            }
+        }
+        break;
+    }
+}
```

Comment:
It seems to me that this function assumes compiler internals that should not be in the DIP (could be in an appendix). Instead a rule set should be given (which the function might satisfy).

Reply:
What are the compiler internals that it assumes? It looks like compiler code, but the entities used (data structures, naming etc.) are actually abstract.

What I tried to avoid with the rule set is ending up with a DIP that will be too abstract and in essence, would just say "fix the thing". I might be wrong and that could be done better. But what I tried to do is to encode (abstractly, as this is not code) how cases have to be handled.
The rules that the function has to satisfy are my explanation on what the "code" handles.
I don't know if this made any sense. If not, please let me know to explain it more. Essentially, the rules are already there. Maybe not in the best way, but _what_ has to be handled is already there (the rules) and to my thinking, the encoding is part of the rules. Because, how do you describe the rule without providing an encoding?